### PR TITLE
DEPR (string): non-bool na for obj.str.contains

### LINKED ISSUE #59561: BUG: .str.contains `na` validation
```python
import pandas as pd
import pyarrow as pa

ser = pd.Series(['a', 'b', None], dtype=pd.StringDtype(storage="pyarrow"))
ser2 = ser.astype(pd.ArrowDtype(pa.string()))

ser.str.contains("foo", na="bar")  # <- casts "bar" to True
ser2.str.contains("foo", na="bar")  # <- raises
```

There's a small difference in the _str_contains methods in ArrowExtensionArray vs ArrowStringArray.  The latter uses `bool(na)` when filling null entries, the former uses `na` directly.

I prefer the no-casting behavior, but mainly think we should be consistent.

<b>update</b> Looks like `pandas/tests/strings/test_find_replace.py::test_contains_nan` specifically tests `na="foo"`

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #59561 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

No 2.3 whatsnew file yet.  Not sure if the plan is to introduce one in #59513

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
I think startswith/endswith have the same issue? If so, can you add the same deprecation message there as well?

> No 2.3 whatsnew file yet. Not sure if the plan is to introduce one in https://github.com/pandas-dev/pandas/pull/59513

Yeah, we should start a 2.3 whatsnew file. We should add it to main, though (not in the backport PR), so things can be added to that file in PRs like this one.

**Comment by jbrockmendel** [comment]
> I think startswith/endswith have the same issue? If so, can you add the same deprecation message there as well?

Sure, updated.

**Comment by jorisvandenbossche** [comment]
Actually, you could maybe move the warning to core/string/accessors.py, then you don't have to repeat it for the various implementations, and then also ArrowDtype(string) will get it (assuming we want to be consistent and also deprecate it for ArrowDtype, and not only for StringDtype?)

**Comment by jorisvandenbossche** [comment]
> assuming we want to be consistent and also deprecate it for ArrowDtype, and not only for StringDtype?

Of course, that one already errors :)

**Comment by jbrockmendel** [comment]
> then you don't have to repeat it for the various implementations

Because it already errors for the ArrowEA version (but with a pyarrow exception), patching it in the StringAccessor method would be bulky.  I think the thing to do is wait until the deprecation is enforced, then do the validation in the StringAccessor method and get consistent exception/message.

**Comment by jorisvandenbossche** [comment]
> Because it already errors for the ArrowEA version (but with a pyarrow exception), patching it in the StringAccessor method would be bulky. I think the thing to do is wait until the deprecation is enforced, then do the validation in the StringAccessor method and get consistent exception/message.

Sounds good!

File: pandas/tests/strings/test_find_replace.py [hunk 1]
```diff
@@ -254,6 +268,34 @@ def test_contains_nan(any_string_dtype):
...
+        isinstance(dtype, pd.StringDtype) and dtype.storage == "python"
+    ) or dtype == np.dtype("object"):
+        msg = "Allowing a non-bool 'na' in obj.str.startswith is deprecated"
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            ser.str.startswith("kapow", na="baz")
+        msg = "Allowing a non-bool 'na' in obj.str.endswith is deprecated"
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            ser.str.endswith("bar", na="baz")
+    else:
+        # TODO: don't surface pyarrow errors
```
Comment (jorisvandenbossche) [inline pandas/tests/strings/test_find_replace.py:289]:
```suggestion
        # TODO(infer_string): don't surface pyarrow errors
```

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by rhshadrach** [comment]
We want to enforce this in 3.0, correct @jbrockmendel @jorisvandenbossche @mroeschke? Can take it up if so.

**Pull request merged**